src/ngraph/runtime/cpu/pass/cpu_mat_fusion.cpp

```diff
@@ -199,3 +201,104 @@ bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Functi
     }
     return modified;
 }
+
+#define TI(x) std::type_index(typeid(x))
+
+std::shared_ptr<Node> fuse_batch_dot(const std::shared_ptr<Node>& n)
+{
+    auto reshape_pred = [](const std::shared_ptr<Node>& r) {
+        return std::dynamic_pointer_cast<op::Reshape>(r) != nullptr;
+    };
+
+    const int num_op_branches = 2;
+    std::shared_ptr<pattern::op::Label> input[num_op_branches];
+    std::shared_ptr<op::Reshape> reshape[num_op_branches];
+    for (int i = 0; i < num_op_branches; ++i)
+    {
+        input[i] = std::make_shared<pattern::op::Label>(element::f32, Shape{3, 2, 2});
+        auto slice =
+            std::make_shared<op::Slice>(input[i], Coordinate{0, 0, 0}, Coordinate{1, 2, 2});
+        auto skip = std::make_shared<pattern::op::Skip>(slice, reshape_pred);
+        reshape[i] = std::make_shared<op::Reshape>(skip, AxisVector{0, 1, 2}, Shape{2, 2});
+    }
+    auto dot = std::make_shared<op::Dot>(reshape[0], reshape[1]);
+    auto final_reshape = std::make_shared<op::Reshape>(dot, AxisVector{0, 1}, Shape{1, 2, 2});
+
+    auto matcher = std::make_shared<pattern::Matcher>(final_reshape);
+    std::shared_ptr<Node> fuse_input[num_op_branches];
+    bool transpose[num_op_branches] = {false, false};
+    const int num_expected_reshape_with_trans = 3;
+
+    // check each input arg matches the pattern
+    for (auto arg : n->get_arguments())
+    {
+        if (matcher->match(arg))
+        {
+            auto pattern_map = matcher->get_pattern_map();
+            int reshape_count[num_op_branches] = {0, 0};
+            // we found a match, determine whether we have to transpose for each input by
+            // counting the number of reshapes in each branch, if transpose is applied, there
+            // should be 3 reshapes.
+            for (int i = 0; i < num_op_branches; ++i)
+            {
+                auto iter = matcher->get_match_root();
+                auto& input_node = pattern_map[input[i]];
+                do
+                {
+                    if (std::dynamic_pointer_cast<op::Reshape>(iter) != nullptr)
```

Comment:
do we want to check what those reshapes actually do (i.e. transpose or reshape or god forbids both at the same time) ?